src/V3Force.cpp

```diff
@@ -277,45 +291,120 @@ class ForceConvertVisitor final : public VNVisitor {
-                    break;
-                case VAccess::WRITE:
-                    // Write references use the original signal
-                    break;
-                default:
-                    nodep->v3error(
-                        "Unsupported: Signals used via read-write reference cannot be forced");
-                    break;
-                }
+                FileLine* const flp = nodep->fileline();
+                AstVarRef* const origp = new AstVarRef{flp, nodep->varScopep(), VAccess::READ};
+                markNonReplaceable(origp);
+                nodep->varp(fcp->m_rdVscp->varp());
+                nodep->varScopep(fcp->m_rdVscp);
             }
-        });
+            break;
+        }
+        case VAccess::WRITE: {
+            if (!m_inLogic) return;
+            // Emit rdVscp update after each write to any VarRef on forced LHS.
+            if (ForceComponentsVarScope* const fcp
+                = m_forceComponentsVarScope.tryGet(nodep->varScopep())) {
+                FileLine* const flp = nodep->fileline();
+                AstVarRef* const lhsp = new AstVarRef{flp, fcp->m_rdVscp, VAccess::WRITE};
+                AstNodeExpr* const rhsp = fcp->forcedUpdate(nodep->varScopep());
+                m_stmtp->addNextHere(new AstAssign{flp, lhsp, rhsp});
+            }
+            // Emit valVscp update after each write to any VarRef on forced RHS.
+            if (AstVarScope* valVscp = getValVscp(nodep)) {
+                FileLine* const flp = nodep->fileline();
+                AstVarRef* const valp = new AstVarRef{flp, valVscp, VAccess::WRITE};
+                AstVarRef* const rhsp = new AstVarRef{flp, nodep->varScopep(), VAccess::READ};
+
+                markNonReplaceable(valp);
+                markNonReplaceable(rhsp);
+
+                m_stmtp->addNextHere(new AstAssign{flp, valp, rhsp});
+            }
+            break;
+        }
+        default:
+            nodep->v3error("Unsupported: Signals used via read-write reference cannot be forced");
+            break;
+        }
     }
+    void visit(AstNode* nodep) override { iterateChildren(nodep); }
 
 public:
-    static void apply(AstNetlist* nodep) { ForceConvertVisitor{nodep}; }
+    // CONSTRUCTOR
+    explicit ForceReplaceVisitor(
+        AstNetlist* nodep,
+        AstUser1Allocator<AstVarScope, ForceComponentsVarScope>& forceComponentsVarScope)
+        : m_forceComponentsVarScope{forceComponentsVarScope} {
+        iterateChildren(nodep);
+    }
 };
-
 //######################################################################
 //
 
 void V3Force::forceAll(AstNetlist* nodep) {
     UINFO(2, __FUNCTION__ << ": " << endl);
     if (!v3Global.hasForceableSignals()) return;
-    ForceConvertVisitor::apply(nodep);
+    {
+        const VNUser1InUse m_user1InUse;
+        const VNUser2InUse m_user2InUse;
+        const VNUser3InUse m_user3InUse;
+        AstUser1Allocator<AstVarScope, ForceComponentsVarScope> forceComponentsVarScope;
```

Comment:
Prefer make a ForceBase class which has this, and both visitors reference that class (see other .cpp files for examples).  Then your user2 static functions can move inside it also.